### PR TITLE
[Merged by Bors] - feat: `expect_...` tactics for testing

### PR DESCRIPTION
**Pull request by arthurpaulino**
feat: `expect_failure`, `expect_failure_message`, `expect_goal` and `expect_hyp`

### CODE REVIEW

File: Mathlib/Tactic/Expect.lean [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+-/
+
+import Lean
+
+namespace Mathlib.Tactic
+
+open Lean Meta Elab Tactic
+
+/-- `expect_failure tacs` succeeds iff `tacs` fails. -/
+elab "expect_failure " tacs:tacticSeq : tactic => withMainContext do
```
Comment (semorrison) [inline Mathlib/Tactic/Expect.lean:14]: Is `withMainContext` needed for `expect_failure`/`expect_failure_msg`/`expect_goal`? It seems maybe not; I would like to understand better when it is needed / desirable!

(It is needed for `expect_hyp`.)

**Comment by semorrison** [comment]
What is your plan for how these overlap with `guard_target` and `guard_hyp`?

File: Mathlib/Tactic/Expect.lean (hunk 1 above)
Comment (arthurpaulino) [reply Mathlib/Tactic/Expect.lean]: Nice catch!
I don't have a complete understanding either. Maybe @digama0 or @gebner can help us here

**Comment by arthurpaulino** [comment]
@semorrison To be honest, I wasn't aware of `guard_hyp` and `guard_target`.

But since you've asked me, I'd be in favor of keeping the `expect_hyp` and `expect_goal` tactics and dropping the others for the sake of naming standardization (along with the other `expect_...` tactics).

**Comment by semorrison** [comment]
Can you check if `guard_hyp` and `guard_target` are used in mathlib outside the tests? I think we should ask @digama0 what he prefers here. Probably the right answer is to keep the existing ones, with the existing names, but possibly add a TODO in a comment suggesting we regularize all the names once the port is done.

**Comment by semorrison** [comment]
Actually, @arthurpaulino, how about just remove `expect_hyp` and `expect_goal` for now. It would be good to get this merged sooner rather than later, as PRs like #318 need `expect_failure` as an alternative to `fail_if_success`.

File: Mathlib/Tactic/Expect.lean [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    throwError "too many expected goals"
+  typesList.zip goalsList |>.forM fun (type, mvarId) => do
+    let goalType ← instantiateMVars (← getMVarDecl mvarId).type
+    let expectedType ← elabTerm type none
+    if goalType != expectedType then
+      throwErrorAt type s!"expected '{expectedType}' but got '{goalType}'"
+
+syntax hypRule := ident " : " term
+
+/-- `expect_hyps h₁ : t₁, h₂ : t₂, ⋯` succeeds iff every `hᵢ` in the context has type `tᵢ`. -/
```
Comment (semorrison) [inline Mathlib/Tactic/Expect.lean:43]: Similarly for `guard_hyp`.

Comment (semorrison) [inline Mathlib/Tactic/Expect.lean:43]:
```suggestion
/-- `expect_hyps h₁ : t₁, h₂ : t₂, ⋯` succeeds iff each `hᵢ` in appears in the context with type `tᵢ`. -/
```

File: Mathlib/Tactic/Expect.lean [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+Authors: Arthur Paulino
+-/
+
+import Lean
+
+namespace Mathlib.Tactic
+
+open Lean Meta Elab Tactic
+
+/-- `expect_failure tacs` succeeds iff `tacs` fails. -/
```
[the remark this answers is not in the document]
Comment (arthurpaulino) [reply Mathlib/Tactic/Expect.lean:13]: I believe they will change `success_if_failure` in core soon.

[the remark this answers is not in the document]
Comment (semorrison) [reply Mathlib/Tactic/Expect.lean:13]: That could go in the comment. :-)

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**